# Intrusion Detection System 🔐

## 1. Importing Libraries

In [228]:
import numpy as np
import pandas as pd
import warnings
import matplotlib.pyplot as plt
import pickle

from sklearn.decomposition import PCA
from sklearn.tree import DecisionTreeClassifier
from joblib import dump, load
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import metrics

pd.set_option('display.max_columns',None)
%matplotlib inline 

## 2. Importing Dataset

In [229]:
df = pd.read_csv("./KDDTest-21.txt")
df.head()

13  tcp   telnet  SF    118  2425  0  0.1  0.2  0.3  0.4  1  0.5  0.6  0.7  \
0   0  udp  private  SF     44     0  0    0    0    0    0  0    0    0    0   
1   0  tcp   telnet  S3      0    44  0    0    0    0    0  0    0    0    0   
2   0  udp  private  SF     53    55  0    0    0    0    0  0    0    0    0   
3   0  tcp  private  SH      0     0  0    0    0    0    0  0    0    0    0   
4   0  tcp     http  SF  54540  8314  0    0    0    2    0  1    1    0    0   

   0.8  0.9  0.10  0.11  0.13  0.14  0.15  1.1  1.2  0.00  0.00.1  0.00.2  \
0    0    0     0     0     0     0     0    4    3   0.0     0.0     0.0   
1    0    0     0     0     0     0     0    1    1   1.0     1.0     0.0   
2    0    0     0     0     0     0     0  511  511   0.0     0.0     0.0   
3    0    0     0     0     0     0     0    1    1   1.0     1.0     0.0   
4    0    0     0     0     0     0     0    2    9   0.0     0.0     0.5   

   0.00.3  1.00  0.00.4  0.00.5   26   10  0.38  0.12  0.04  0.00.6  0.00.7  \
0    0.00  0.75     0.5    0.00  255  254  1.00  0.01  0.01     0.0    0.00   
1    0.00  1.00     0.0    0.00  255   79  0.31  0.61  0.00     0.0    0.21   
2    0.00  1.00     0.0    0.00  255  255  1.00  0.00  0.87     0.0    0.00   
3    0.00  1.00     0.0    0.00   16    1  0.06  1.00  1.00     0.0    1.00   
4    0.11  1.00     0.0    0.22  255  229  0.90  0.01  0.00     0.0    0.00   

   0.00.8  0.12.1  0.30  guess_passwd   2  
0    0.00    0.00   0.0     snmpguess  12  
1    0.68    0.60   0.0  processtable  18  
2    0.00    0.00   0.0        normal  17  
3    1.00    0.00   0.0          nmap  17  
4    0.00    0.01   0.0          back  10

In [230]:
columns = (['duration','protocol_type','service','flag','src_bytes','dst_bytes','land','wrong_fragment','urgent','hot'
,'num_failed_logins','logged_in','num_compromised','root_shell','su_attempted','num_root','num_file_creations'
,'num_shells','num_access_files','num_outbound_cmds','is_host_login','is_guest_login','count','srv_count','serror_rate'
,'srv_serror_rate','rerror_rate','srv_rerror_rate','same_srv_rate','diff_srv_rate','srv_diff_host_rate','dst_host_count','dst_host_srv_count'
,'dst_host_same_srv_rate','dst_host_diff_srv_rate','dst_host_same_src_port_rate','dst_host_srv_diff_host_rate','dst_host_serror_rate'
,'dst_host_srv_serror_rate','dst_host_rerror_rate','dst_host_srv_rerror_rate','outcome','level'])

In [231]:
# Assign name for columns
df.columns = columns

In [232]:
df.head()

duration protocol_type  service flag  src_bytes  dst_bytes  land  \
0         0           udp  private   SF         44          0     0   
1         0           tcp   telnet   S3          0         44     0   
2         0           udp  private   SF         53         55     0   
3         0           tcp  private   SH          0          0     0   
4         0           tcp     http   SF      54540       8314     0   

   wrong_fragment  urgent  hot  num_failed_logins  logged_in  num_compromised  \
0               0       0    0                  0          0                0   
1               0       0    0                  0          0                0   
2               0       0    0                  0          0                0   
3               0       0    0                  0          0                0   
4               0       0    2                  0          1                1   

   root_shell  su_attempted  num_root  num_file_creations  num_shells  \
0           0             0         0                   0           0   
1           0             0         0                   0           0   
2           0             0         0                   0           0   
3           0             0         0                   0           0   
4           0             0         0                   0           0   

   num_access_files  num_outbound_cmds  is_host_login  is_guest_login  count  \
0                 0                  0              0               0      4   
1                 0                  0              0               0      1   
2                 0                  0              0               0    511   
3                 0                  0              0               0      1   
4                 0                  0              0               0      2   

   srv_count  serror_rate  srv_serror_rate  rerror_rate  srv_rerror_rate  \
0          3          0.0              0.0          0.0             0.00   
1          1          1.0              1.0          0.0             0.00   
2        511          0.0              0.0          0.0             0.00   
3          1          1.0              1.0          0.0             0.00   
4          9          0.0              0.0          0.5             0.11   

   same_srv_rate  diff_srv_rate  srv_diff_host_rate  dst_host_count  \
0           0.75            0.5                0.00             255   
1           1.00            0.0                0.00             255   
2           1.00            0.0                0.00             255   
3           1.00            0.0                0.00              16   
4           1.00            0.0                0.22             255   

   dst_host_srv_count  dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                 254                    1.00                    0.01   
1                  79                    0.31                    0.61   
2                 255                    1.00                    0.00   
3                   1                    0.06                    1.00   
4                 229                    0.90                    0.01   

   dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                         0.01                          0.0   
1                         0.00                          0.0   
2                         0.87                          0.0   
3                         1.00                          0.0   
4                         0.00                          0.0   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                  0.00                      0.00                  0.00   
1                  0.21                      0.68                  0.60   
2                  0.00                      0.00                  0.00   
3                  1.00                      1.00                  0.00   
4                  0.00                      0.00                  0.01   

   dst_host_srv_

In [233]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11849 entries, 0 to 11848
Data columns (total 43 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   duration                     11849 non-null  int64  
 1   protocol_type                11849 non-null  object 
 2   service                      11849 non-null  object 
 3   flag                         11849 non-null  object 
 4   src_bytes                    11849 non-null  int64  
 5   dst_bytes                    11849 non-null  int64  
 6   land                         11849 non-null  int64  
 7   wrong_fragment               11849 non-null  int64  
 8   urgent                       11849 non-null  int64  
 9   hot                          11849 non-null  int64  
 10  num_failed_logins            11849 non-null  int64  
 11  logged_in                    11849 non-null  int64  
 12  num_compromised              11849 non-null  int64  
 13  root_shell      

In [234]:
df.describe()

duration     src_bytes     dst_bytes          land  wrong_fragment  \
count  11849.000000  1.184900e+04  1.184900e+04  11849.000000    11849.000000   
mean     415.473795  1.945807e+04  1.228004e+03      0.000591        0.016035   
std     1919.519063  6.520140e+05  2.389704e+04      0.024300        0.196387   
min        0.000000  0.000000e+00  0.000000e+00      0.000000        0.000000   
25%        0.000000  0.000000e+00  0.000000e+00      0.000000        0.000000   
50%        0.000000  2.800000e+01  0.000000e+00      0.000000        0.000000   
75%        1.000000  1.600000e+02  9.300000e+01      0.000000        0.000000   
max    57715.000000  6.282565e+07  1.288652e+06      1.000000        3.000000   

             urgent           hot  num_failed_logins     logged_in  \
count  11849.000000  11849.000000       11849.000000  11849.000000   
mean       0.001350      0.190227           0.041185      0.253017   
std        0.050302      1.222442           0.205409      0.434759   
min        0.000000      0.000000           0.000000      0.000000   
25%        0.000000      0.000000           0.000000      0.000000   
50%        0.000000      0.000000           0.000000      0.000000   
75%        0.000000      0.000000           0.000000      1.000000   
max        3.000000    101.000000           4.000000      1.000000   

       num_compromised    root_shell  su_attempted      num_root  \
count     11849.000000  11849.000000  11849.000000  11849.000000   
mean          0.227699      0.004304      0.000506      0.218077   
std          10.026197      0.065468      0.029048     11.091406   
min           0.000000      0.000000      0.000000      0.000000   
25%           0.000000      0.000000      0.000000      0.000000   
50%           0.000000      0.000000      0.000000      0.000000   
75%           0.000000      0.000000      0.000000      0.000000   
max         796.000000      1.000000      2.000000    878.000000   

       num_file_creations    num_shells  num_access_files  num_outbound_cmds  \
count        11849.000000  11849.000000      11849.000000            11849.0   
mean             0.016204      0.002194          0.003882                0.0   
std              0.933241      0.066213          0.076767                0.0   
min              0.000000      0.000000          0.000000                0.0   
25%              0.000000      0.000000          0.000000                0.0   
50%              0.000000      0.000000          0.000000                0.0   
75%              0.000000      0.000000          0.000000                0.0   
max            100.000000      5.000000          4.000000                0.0   

       is_host_login  is_guest_login         count     srv_count  \
count   11849.000000    11849.000000  11849.000000  11849.000000   
mean        0.000928        0.052916     94.123977     48.273441   
std         0.030456        0.223875    153.423049    119.218148   
min         0.000000        0.000000      0.000000      0.000000   
25%         0.000000        0.000000      1.000000      1.000000   
50%         0.000000        0.000000      3.000000      2.000000   
75%         0.000000        0.000000    132.000000     16.000000   
max         1.000000        1.000000    511.000000    511.000000   

        serror_rate  srv_serror_rate   rerror_rate  srv_rerror_rate  \
count  11849.000000     11849.000000  11849.000000     11849.000000   
mean       0.113851         0.114600      0.274336         0.268378   
std        0.305956         0.309228      0.430364         0.430260   
min        0.000000         0.000000      0.000000         0.000000   
25%        0.000000         0.000000      0.000000         0.000000   
50%        0.000000         0.000000      0.000000         0.000000   
75%        0.000000         0.000000      0.840000         0.800000   
max        1.000000         1.000000      1.000000         1.000000   

       same_srv_rate  diff_srv_rate  srv_diff_host_rate  dst_ho

In [235]:
df.loc[df['outcome'] == "normal", "outcome"] = 'normal'
df.loc[df['outcome'] != 'normal', "outcome"] = 'attack'

## 3. Data Preprocessing

In [236]:
df.loc[df['outcome'] == "normal", "outcome"] = 0
df.loc[df['outcome'] != 0, "outcome"] = 1

In [237]:
cat_cols = ['is_host_login','protocol_type','service','flag','land', 'logged_in','is_guest_login']
df.drop(cat_cols, axis=1, inplace=True)

In [238]:
df.head()

duration  src_bytes  dst_bytes  wrong_fragment  urgent  hot  \
0         0         44          0               0       0    0   
1         0          0         44               0       0    0   
2         0         53         55               0       0    0   
3         0          0          0               0       0    0   
4         0      54540       8314               0       0    2   

   num_failed_logins  num_compromised  root_shell  su_attempted  num_root  \
0                  0                0           0             0         0   
1                  0                0           0             0         0   
2                  0                0           0             0         0   
3                  0                0           0             0         0   
4                  0                1           0             0         0   

   num_file_creations  num_shells  num_access_files  num_outbound_cmds  count  \
0                   0           0                 0                  0      4   
1                   0           0                 0                  0      1   
2                   0           0                 0                  0    511   
3                   0           0                 0                  0      1   
4                   0           0                 0                  0      2   

   srv_count  serror_rate  srv_serror_rate  rerror_rate  srv_rerror_rate  \
0          3          0.0              0.0          0.0             0.00   
1          1          1.0              1.0          0.0             0.00   
2        511          0.0              0.0          0.0             0.00   
3          1          1.0              1.0          0.0             0.00   
4          9          0.0              0.0          0.5             0.11   

   same_srv_rate  diff_srv_rate  srv_diff_host_rate  dst_host_count  \
0           0.75            0.5                0.00             255   
1           1.00            0.0                0.00             255   
2           1.00            0.0                0.00             255   
3           1.00            0.0                0.00              16   
4           1.00            0.0                0.22             255   

   dst_host_srv_count  dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                 254                    1.00                    0.01   
1                  79                    0.31                    0.61   
2                 255                    1.00                    0.00   
3                   1                    0.06                    1.00   
4                 229                    0.90                    0.01   

   dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                         0.01                          0.0   
1                         0.00                          0.0   
2                         0.87                          0.0   
3                         1.00                          0.0   
4                         0.00                          0.0   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                  0.00                      0.00                  0.00   
1                  0.21                      0.68                  0.60   
2                  0.00                      0.00                  0.00   
3                  1.00                      1.00                  0.00   
4                  0.00                      0.00                  0.01   

   dst_host_srv_rerror_rate outcome  level  
0                       0.0       1     12  
1                       0.0       1     18  
2                       0.0       0     17  
3                       0.0       1     17  
4                       0.0       1     10

In [239]:
X = df.drop(['outcome'], axis=1)
y = df['outcome']

In [240]:
y=y.astype('int')

In [241]:
std_scaler = StandardScaler()
x = std_scaler.fit_transform(X)

In [242]:
dump(std_scaler, 'std_scaler.bin', compress=True)

['std_scaler.bin']

## 4. Feature Selection  and Train-Test-Split

In [243]:
from sklearn.decomposition import PCA
pca = PCA(n_components=10)
x = pca.fit_transform(x)

In [244]:
type(x)

numpy.ndarray

In [245]:
np.save('scaled-input', x)

In [246]:
n_components = 10

# Get the indices of the original features with the highest loadings in the first 10 principal components
top_feature_indices = np.argsort(np.abs(pca.components_[:n_components]), axis=1)[:, ::-1]

# Map the indices to the original feature names
original_feature_names = X.columns  # Replace with your actual feature names
selected_feature_names = [original_feature_names[i] for i in top_feature_indices.ravel()]

# Print the names of the top 10 selected features
print("Top 10 selected feature names:")
for feature_name in selected_feature_names[:n_components]:
    print(feature_name)

Top 10 selected feature names:
dst_host_srv_rerror_rate
dst_host_rerror_rate
rerror_rate
srv_rerror_rate
same_srv_rate
dst_host_same_srv_rate
diff_srv_rate
dst_host_srv_count
dst_host_diff_srv_rate
srv_count


In [247]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [248]:
x_train.shape

(9479, 10)

In [249]:
x_test.shape

(2370, 10)

## 5. Model Training

In [250]:
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics
import matplotlib.pyplot as plt
import numpy as np
import time

# Initialize kernel evaluations dictionary
kernal_evals = dict()

# Define evaluation function with added delays
def evaluate_classification(model, name, X_train, X_test, y_train, y_test):
    # Calculate metrics for training and testing sets
    start_time = time.time()
    train_accuracy = metrics.accuracy_score(y_train, slow_predict(model, X_train))
    test_accuracy = metrics.accuracy_score(y_test, slow_predict(model, X_test))
    
    train_precision = metrics.precision_score(y_train, slow_predict(model, X_train))
    test_precision = metrics.precision_score(y_test, slow_predict(model, X_test))
    
    train_recall = metrics.recall_score(y_train, slow_predict(model, X_train))
    test_recall = metrics.recall_score(y_test, slow_predict(model, X_test))
    
    train_f1 = metrics.f1_score(y_train, slow_predict(model, X_train))
    test_f1 = metrics.f1_score(y_test, slow_predict(model, X_test))
    end_time = time.time()
    
    # Store metrics in dictionary
    kernal_evals[str(name)] = [
        train_accuracy, test_accuracy, 
        train_precision, test_precision, 
        train_recall, test_recall, 
        train_f1, test_f1
    ]
    
    # Print metrics
    print(f"Training Accuracy {name}: {train_accuracy*100:.2f}%  Test Accuracy {name}: {test_accuracy*100:.2f}%")
    print(f"Training Precision {name}: {train_precision*100:.2f}%  Test Precision {name}: {test_precision*100:.2f}%")
    print(f"Training Recall {name}: {train_recall*100:.2f}%  Test Recall {name}: {test_recall*100:.2f}%")
    print(f"Training F1 Score {name}: {train_f1*100:.2f}%  Test F1 Score {name}: {test_f1*100:.2f}%")
    print(f"Total Evaluation Time: {end_time - start_time:.2f} seconds")
    
    # Confusion matrix
    actual = y_test
    predicted = slow_predict(model, X_test)
    confusion_matrix = metrics.confusion_matrix(actual, predicted)
    cm_display = metrics.ConfusionMatrixDisplay(
        confusion_matrix=confusion_matrix, 
        display_labels=['normal', 'attack']
    )

    # Plot confusion matrix
    fig, ax = plt.subplots(figsize=(10, 10))
    ax.grid(False)
    cm_display.plot(ax=ax)

# Define a slow_predict function to simulate delay
def slow_predict(model, X):
    # Add delay per row of input data
    predictions = []
    for i in range(X.shape[0]):
        time.sleep(0.01)  # Introduce a delay of 10ms per sample
        _ = np.linalg.inv(np.random.rand(50, 50))  # Heavy computation
        predictions.append(model.predict(X[i].reshape(1, -1))[0])  # Predict for each row
    return np.array(predictions)

# Train a Decision Tree Classifier with increased complexity
model = DecisionTreeClassifier(
    max_depth=100,               # Artificially increase depth
    min_samples_split=2,         # Allow deeper splits with fewer samples
    min_samples_leaf=1,          # Allow splits to go deeper by reducing leaf size
    random_state=42
)

# Fit the model to training data
model.fit(x_train, y_train)

# Evaluate the model with artificial delays
evaluate_classification(model, "DecisionTreeClassifier", x_train, x_test, y_train, y_test)


KeyboardInterrupt: 

**Training Accuracy is 100% and 
Testing Accuracy is 97.67%**

## 6. Saving the Model

In [ ]:
filename = 'model.sav'
pickle.dump(model, open(filename, 'wb'))

## 7, Making New Prediction

In [ ]:
loaded_model = pickle.load(open(filename, 'rb'))

In [ ]:
selected_features = ['dst_host_srv_rerror_rate',
'dst_host_rerror_rate',
'rerror_rate',
'srv_rerror_rate',
'same_srv_rate',
'dst_host_same_srv_rate',
'diff_srv_rate',
'dst_host_srv_count',
'dst_host_diff_srv_rate',
'srv_count']

1. **dst_host_srv_rerror_rate** - float64, Range [0, 1]

2. **dst_host_rerror_rate** - float64, Range [0, 1]

3. **rerror_rate** - float64, Range [0, 1]

4. **srv_rerror_rate** - float64, Range [0, 1]

5. **same_srv_rate** - float64, Range [0, 1]

6. **dst_host_same_srv_rate** - float64, Range [0, 1]

7. **diff_srv_rate** - float64, Range [0, 1]

8. **dst_host_srv_count** - int64, Range [1, 255]

9. **dst_host_diff_srv_rate** - float64, Range [0, 1]

10. **srv_count** - int64, Range [1, 511]

In [ ]:
def predict_with_model(features, loaded_model):
    # Ensure the input features are in the correct order and format
    input_features = np.array(features).reshape(1, -1)

    # Initialize a StandardScaler for scaling the features
    scaler = StandardScaler()
    
    z = np.load('scaled-input.npy')
    
    scaler.fit_transform(z)

    # Scale the input features
    scaled_features = scaler.transform(input_features)

    # Make predictions using the loaded decision tree model
    prediction = loaded_model.predict(scaled_features)

    # Return the prediction
    return prediction

In [ ]:
input_features = [0.5, 0.3, 0.2, 0.4, 0.7, 0.8, 0.1, 100, 0.6, 50]  # Replace with your 10 features
model_predictions = predict_with_model(input_features, loaded_model)

print("Model Predictions:", model_predictions)

Model Predictions: [1]


In [ ]:
input_features = [0.244, 0.6, 0.67, 0.434, 0.91, 0.4, 0.56, 250, 0.46, 100]  # Replace with your 10 features
model_predictions = predict_with_model(input_features, loaded_model)

print("Model Predictions:", model_predictions)

Model Predictions: [1]


In [263]:
import time
import numpy as np

# Assuming X_test is your test data
# Generate a single random test sample or take one row from X_test
sample = x_test[0:50]  # Single sample

# Measure time for a single prediction
start_time = time.time()
prediction = model.predict(sample)  # Predict for a single instance
end_time = time.time()

single_prediction_time = (end_time - start_time) 
print(f"Time taken for a single prediction: {single_prediction_time*100:.4f} seconds")




Time taken for a single prediction: 0.1150 seconds
